GeosUtil/unitconv_mod.F90

```diff
@@ -2467,31 +2536,173 @@ SUBROUTINE ConvertBox_Kg_to_Kgm2( I, J, L, State_Chm, State_Grid, isAdjoint, RC
+       ENDIF
 #endif
+
+       ! Update units metadata
+       State_Chm%Species(N)%Previous_Units = State_Chm%Species(N)%Units
+       State_Chm%Species(N)%Units = KG_SPECIES_PER_M2
+
     ENDDO
     !$OMP END PARALLEL DO
 
   END SUBROUTINE ConvertBox_Kg_to_Kgm2
 !EOC
+!------------------------------------------------------------------------------
+!                  GEOS-Chem Global Chemical Transport Model                  !
+!------------------------------------------------------------------------------
+!BOP
+!
+! !IROUTINE: Check_Units
+!
+! !DESCRIPTION: Ensures that all species have the same Units value before
+!  unit conversion is done.
+!\\
+! !INTERFACE:
+!
+  SUBROUTINE Check_Units( State_Chm, mapping, units, RC )
+!
+! !INPUT PARAMETERS:
+!
+    INTEGER,        INTENT(IN)    :: mapping(:)  ! Mapping to species Id
+    INTEGER,        INTENT(IN)    :: units       ! Input units flag
+!
+! !INPUT/OUTPUT PARAMETERS:
+!
+    TYPE(ChmState), INTENT(INOUT) :: State_Chm   ! Chemistry State object
+!
+! !OUTPUT PARAMETERS:
+!
+    INTEGER,        INTENT(OUT)   :: RC          ! Success or failure
+!
+! !REVISION HISTORY:
+!  30 Nov 2023 - R. Yantosca - Initial version
+!  See the subsequent Git history with the gitk browser!
+!EOP
+!------------------------------------------------------------------------------
+!BOC
+!
+! !LOCAL VARIABLES:
+!
+    ! Scalars
+    INTEGER            :: N,      S
+
+    ! Strings
+    CHARACTER(LEN=255) :: errMsg, thisLoc
+
+    !========================================================================
+    ! Check_Units begins here!
+    !========================================================================
+
+    ! Initialize
+    RC      = GC_SUCCESS
+    errMsg  = ''
+    thisLoc = ' -> at Check_Units (in module GeosUtil/unitconv_mod.F90)'
+
+    ! Make sure all species start with the proper unit, or throw an error
+    IF ( .not. ALL( State_Chm%Species(mapping)%Units == units ) ) THEN
+       errMsg = 'All species do not have Units = '                        // &
+                 TRIM( UNIT_STR( units ) )
+       CALL GC_Error( errMsg, RC, thisLoc )
+       RETURN
+    ENDIF
+
+  END SUBROUTINE Check_Units
+!EOC
+!------------------------------------------------------------------------------
+!                  GEOS-Chem Global Chemical Transport Model                  !
+!------------------------------------------------------------------------------
+!BOP
+!
+! !IROUTINE: Check_Previous_Units
+!
+! !DESCRIPTION:  Ensures that all species have the same Previous_Units value
+!  before unit conversion is done.
+!\\
+!\\
+! !INTERFACE:
+!
+  SUBROUTINE Check_Previous_Units( State_Chm, mapping, units, RC )
+!
+! !INPUT PARAMETERS:
+!
+    INTEGER,        INTENT(IN)    :: mapping(:)  ! Mapping to species Id
+    INTEGER,        INTENT(IN)    :: units       ! Input unit flag
+!
+! !INPUT/OUTPUT PARAMETERS:
+!
+    TYPE(ChmState), INTENT(INOUT) :: State_Chm   ! Chemistry State object
```

Comment:
Would there be any benefit to making this a function that returns true/false?

Reply:
We have also removed the `Check_Previous_Units` routine as this is not really used.  It will also save some CPU cycles.